src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -1322,16 +1542,44 @@ private static unsafe nuint NarrowUtf16ToAscii_Sse2(char* pUtf16Buffer, byte* pA
             }
             else
             {
-                if ((Sse2.MoveMask(Sse2.AddSaturate(utf16VectorFirst.AsUInt16(), asciiMaskForAddSaturate).AsByte()) & NonAsciiDataSeenMask) != 0)
+                if (Sse2.IsSupported)
                 {
-                    return 0;
+                    if ((Sse2.MoveMask(Sse2.AddSaturate(utf16VectorFirst.AsUInt16(), asciiMaskForAddSaturate).AsByte()) & NonAsciiDataSeenMask) != 0)
+                    {
+                        return 0;
+                    }
+                }
+                else if (AdvSimd.Arm64.IsSupported)
+                {
+                    if (ContainsNonAsciiValue(AdvSimd.AddSaturate(utf16VectorFirst.AsUInt16(), asciiMaskForAddSaturate)))
```

Comment:
I'd just like a second pair of eyes here: Based on Levi's explanation, this looks correct to me. The result of `AdvSimd.AddSaturate(utf16VectorFirst.AsUInt16(), asciiMaskForAddSaturate))` will have the `0x8000` bit set for non-Ascii values and `ContainsNonAsciiValue` ANDs with `0x8000800080008000`, so the logic is equivalent I think. @kunalspathak

Reply:
Yes, that's my understanding.  The `& NonAsciiDataSeenMask` is needed in Sse2 case because during `MoveMask()` we convert data from `ushort` to `byte`, but to find non-ascii, we need to check the alternate lanes of byte variants of what `MoveMask` returned and hence  we `& 0b_1010_1010_1010_1010`. But for AdvSimd, in `ContainsNonAsciiValue(ushort)` you are already operating on `ushort` and there as per @GrabYourPitchforks explaination, you just need to check if there is any lane whose MSB is set (as a result of `AddSaturate()` ) .

I still need to double check your code around `NonAsciiDataSeenMask` and `GetMaskOfMostSignificantBits`.